### PR TITLE
Change the way renderer might be passed to HotColumn and HotTable to use render props approach

### PR DESCRIPTION
**Pull request by NOtherDev**
### Context

Specifying renderers as children components marked with "magic" `hot-renderer` attribute does not feel idiomatic. It doesn't represent something that is a part of the column, it's purely the column's property. Moreover, this kind of composition allows putting multiple Renderers in one column, which is illogical.

Instead, we're proposing to use a variant of render props pattern and allow specifying component-based renderers via `HotTable` or `HotColumn` props directly. The awesome part of this approach is we can keep things heavily typed in regards to typescript, it is the preferred approach when it comes to proper react and involves less magic behind the scenes.

### How has this been tested?

- All the unit tests still pass + new behavior tests implemented.
- Example projects from the repo still work properly.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. Discussion on RFC: https://github.com/handsontable/handsontable/discussions/10767

### Affected project(s):
- [ ] `handsontable`
- [ ] `@handsontable/angular`
- [x] `@handsontable/react`
- [ ] `@handsontable/vue`
- [ ] `@handsontable/vue3`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [x] My change requires a change to the documentation.

### CODE REVIEW

File: docs/content/guides/getting-started/react-redux.md
Comment (NOtherDev) [inline docs/content/guides/getting-started/react-redux.md:444]: This was an interesting use case, using the same component as both renderer and editor... It wouldn't work in the new approach. I'm not sure if that was something we should fight for, although if this opens a way to share the state between the render and edit stages, maybe it's something we actually overlooked? 🤔

**Comment by NOtherDev** [comment]
@evanSe Due to the complexities of Editors that we might have overlooked in the analysis, I decided to split #3 to cover the Renderers part separately, as it seems to be ready. I'll follow up with a separate PR for Editors once we agree on the final approach.

File: docs/content/guides/getting-started/react-redux.md
Comment (evanSe) [reply docs/content/guides/getting-started/react-redux.md:444]: I was discussing this with a developer recently when I found this 😄, we are fine to ignore this case as its not known to be used by anyone.

File: wrappers/react/src/helpers.tsx
Comment (evanSe) [inline wrappers/react/src/helpers.tsx:53]: Nice one!

File: wrappers/react/src/helpers.tsx
Comment (evanSe) [inline wrappers/react/src/helpers.tsx:168]: Good to see `Math.random()` removed, it was a funny one to find when I was fixing a re-render issue

File: wrappers/react/src/hotTableContext.tsx
Comment (evanSe) [inline wrappers/react/src/hotTableContext.tsx:132]: I do not know if this implementation would have the flicker issue the [pull request](https://github.com/handsontable/handsontable/pull/10758/files) resolves but it would be worth adding the guid at least coming from the instance
```
      // Handsontable.Core type is mising guid
      const instanceGuid = (instance as unknown as { guid: string }).guid;
      const portalContainerKey = `${instanceGuid}${key}`
      const portalKey = `${key}-${instanceGuid}`
```

in this pr the portal container is being cached which resolves a flickering problem, its really noticeable if you have a custom cell renderer with a background color.

File: wrappers/react/src/hotTableContext.tsx
Comment (evanSe) [inline wrappers/react/src/hotTableContext.tsx:150]: We should be able to remove the Math.random() such as `${rowColKey}-${instanceGuid}`

**Review by evanSe — APPROVED** [review]
Already looks so much nicer to use! just some minor comments around including instance.guid

File: wrappers/react/src/hotTableContext.tsx
Comment (NOtherDev) [reply wrappers/react/src/hotTableContext.tsx:132]: I have https://github.com/handsontable/handsontable/pull/10758/files pinned in my TODO list. My idea was to wait until it's merged into `develop`, then switch my development on top of the upstream as well, to have the proper builds running in GH, better visibility etc - I was actually thinking about recreating my PRs on top of upstream, it's just 3 of them by now, I can even do it manually if needed. 

If you agree it makes sense, then I'd not duplicate these changes in my PR and just sync my changes once I'm moving there.

WDYT?

File: wrappers/react/src/hotTableContext.tsx
Comment (NOtherDev) [reply wrappers/react/src/hotTableContext.tsx:150]: As above - I'd be happy to integrate my changes on top of #10758 once it's merged and then `Math.random` will be gone

File: wrappers/react/src/hotTableContext.tsx
Comment (evanSe) [reply wrappers/react/src/hotTableContext.tsx:132]: Makes sense, agreed

**Pull request merged**
